src/hypofuzz/database.py

```diff
@@ -0,0 +1,35 @@
+import json
+from typing import Iterable
+
+from hypothesis import settings
+from hypothesis.database import BackgroundWriteDatabase, ExampleDatabase
+
+
+def metadata_key(key: bytes) -> bytes:
+    return key + b".hypofuzz.metadata"
+
+
+class MetadataDatabase(ExampleDatabase):
+    def __init__(self, db: ExampleDatabase) -> None:
+        self._db = db
+
+    def save(self, key: bytes, value: bytes) -> None:
+        self._db.save(key, value)
+
+    def fetch(self, key: bytes) -> Iterable[bytes]:
+        return self._db.fetch(key)
+
+    def delete(self, key: bytes, value: bytes) -> None:
+        self._db.delete(key, value)
+
+    def save_metadata(self, key: bytes, value: bytes) -> None:
+        self._db.save(metadata_key(key), value)
+
+    def delete_metadata(self, key: bytes, value: bytes) -> None:
+        self._db.delete(metadata_key(key), value)
+
+    def fetch_metadata(self, key: bytes) -> Iterable[bytes]:
+        return [json.loads(v) for v in db.fetch(metadata_key(key))]
+
+
+db = MetadataDatabase(BackgroundWriteDatabase(settings.default.database))
```

Comment:
I think we want a hard cutoff point where we create a database singleton, otherwise we have a thread.start overhead for every db access. But yeah we should be creating the db post-something.

Reply:
Yeah, maybe we wrap this in a `@functools.lru_cache(maxsize=1)` or something?  I agree we only want one per process, it's just a matter of when we create it.